File: app/src/main/java/com/krystianwsul/checkme/gui/tree/ModelNode.java
package com.krystianwsul.checkme.gui.tree;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;

public interface ModelNode extends Comparable<ModelNode> {
    void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder);

    int getItemViewType();

    boolean selectable();

    void onClick();

    boolean visibleWhenEmpty();

    boolean visibleDuringActionMode();

    boolean separatorVisibleWhenNotExapanded();
}

File: treeadapter/src/main/java/com/krystianwsul/treeadapter/ModelNode.kt
package com.krystianwsul.treeadapter

import androidx.recyclerview.widget.RecyclerView

interface ModelNode<T : TreeHolder> : Comparable<ModelNode<T>>, Matchable {

    val itemViewType: Int

    val isSelectable get() = false

    val isSeparatorVisibleWhenNotExpanded get() = false

    val showSeparatorWhenParentExpanded get() = true

    val state: ModelState

    val id: Any

    val propagateSelection get() = false

    val parentNode: ModelNode<T>?

    val inheritParentBottomSeparator get() = false

    fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, startingDrag: Boolean = false)

    fun forceSelected(viewHolder: RecyclerView.ViewHolder): Unit =
        throw UnsupportedOperationException()

    fun onPayload(viewHolder: RecyclerView.ViewHolder): Unit = throw UnsupportedOperationException()

    fun onClick(holder: T) = Unit

    fun tryStartDrag(viewHolder: RecyclerView.ViewHolder): Boolean = false

    fun isVisible(actionMode: Boolean, hasVisibleChildren: Boolean): Boolean = true
}
